Add imdsclient library for querying IMDS

jpculp opened the pull request:
**Issue number:**

N/A

**Description of changes:**

This adds an imdsclient library for tools and setting generators to leverage for fetching user-data, meta-data, and dynamic data from the AWS Instance Metadata Service.

`shibaken` was also updated by removing the stand-alone IMDS logic in favor of the shared imdsclient library.
Some of the log messages in `shibaken` were also tweaked for more clarity.

`pluto` and `early-boot-config` have also been updated by removing the stand-alone IMDS logic in favor of the shared imdsclient library.

**Packages that need to be updated:**
- [x] shibaken
- [x] pluto
- [x] early-boot-config

**Issues:**
- [x] Add testing (https://github.com/bottlerocket-os/bottlerocket/issues/1398)

**Testing done:**

- Built `aws-k8s-1.19` ami and launched instance.
- Instance connected to eks cluster.
- Connected to control container via ssm session.
- Verified that `host-containers.admin.user-data` contained a base64-encoded block.
- Connected to admin container via ssh.
- Verified that `/.bottlerocket/host-containers/admin/user-data` contained JSON.
- Ran `sudo sheltie` to verify root shell was still available.
- Checked for failed systemd units.
- Ran `pluto` with it's sub-commands to verify functionality.

Also built and launched `vmware-k8s-1.20` (x86_64) and `aws-ecs-1` (aarch64).
Verified that they booted correctly and connected to their respective clusters.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

jpculp committed:
Add imdsclient library for querying IMDS

This adds an imdsclient library for tools and setting generators to
leverage for fetching user-data, meta-data, and dynamic data from the
AWS Instance Metadata Service.

jpculp commented:
Seems like the examples in the Doc strings aren't tight enough to pass `Doc-tests`. Will post another revision.

jpculp commented on sources/imdsclient/src/lib.rs:
````diff
@@ -0,0 +1,228 @@
+    /// ```
+    /// # use imdsclient::ImdsClient;
+    /// # let imdsclient=ImdsClient::new().unwrap();
+    /// # let end_target="";
````
Would it improve readability to add some example values to the hidden pieces of the doc string?

zmrow replied:
If we end up moving these examples to the docstring, I would vote we use some "real" values to make them more applicable and run-able by a user.

jpculp replied:
I'm leaning towards removing the examples altogether. The functions themselves are fairly simple and having the examples be Docs-test compatible feels like it's reducing the code readability outside of an IDE.

jpculp commented on sources/api/shibaken/src/main.rs:
```diff
@@ -191,7 +158,7 @@ fn parse_args(args: env::Args) -> Result<Args> {
 
     Ok(Args {
-        log_level: log_level.unwrap_or_else(|| LevelFilter::Info),
+        log_level: log_level.unwrap_or(LevelFilter::Info),
```
Since `sundog` only prints `stderr` in the event of an error, would it make sense for me to change default log level to Debug? It would make it easier to troubleshoot, but might be somewhat ugly if someone ran `shibaken` in a shell.

tjkirch replied:
Do we not have enough information in error/warn/info to get a sense of the problem when it fails?  It seems like we should be adding information there, not just promoting all of the debug messages, if you think there's not enough to understand the problem.

jpculp replied:
It's kind of an edge-case, but the current IMDS logic in Bottlerocket treats a 404 as an `Ok(None)` and that's what we want 99% of the time. However, in `shibaken` we decided to actually error out if it receives a 404 when it is collecting the data from the individual keys in IMDS. Since the `imdsclient` isn't returning an error on 404, some context is lost.

At an error level `shibaken` would return something like:
```
Error retrieving 'public-keys/1/openssh-key' from IMDS metadata
```
but the debug message would show the full uri like this:
```
(2) reqwest::async_impl::client: response '404 Not Found' for http://169.254.169.254/2020-10-27/meta-data/public-keys/1/openssh-key
```

I can update the error to say 404 for clarity, but I'm not quite sure if there is a simple way to pass the full uri. One option is to re-construct the uri via public constants, but I'm not a fan of that since it's not a guarantee that that was the uri that was `reqwest`'d

tjkirch replied:
I don't think the full URI is super helpful, since there's only one thing it could possibly be, given you know the version you're running.  And this case could seemingly only crop up if EC2 listed erroneous keys, or if the user changed the registered keys after the list was retrieved.  I don't think it's possible to change the keys associated in EC2 while an instance is running, so I don't expect that particular issue to come up.  The error seems fine to me...

webern replied:
I think the error message `Error retrieving 'public-keys/1/openssh-key' from IMDS metadata` is clear enough. If you really do have a reproducible case where you need to know whether it was a 404 or something else, then you are really in the weeds anyway and it seems reasonable to expect the user/developer to turn on debugging for something that deep.

zmrow commented on sources/imdsclient/build.rs:
```diff
@@ -0,0 +1,45 @@
+    readme.write_all(content.as_bytes()).unwrap();
+}
+
+fn generate_constants() {
```
Can we add a comment here explaining what this is for?

jpculp replied:
I copied the `build.rs` to be consistent with the `bottlerocket-release` and `parse-datetime` libs, but after looking at it closer and talking with @tjkirch it seems that only `bottlerocket-release` needs this logic and it should be removed from `imdsclient` and `parse-datetime`

zmrow commented on sources/imdsclient/README.md:
```diff
@@ -0,0 +1,13 @@
+Current version: 0.1.0
+
+The imdsclient library provides high-level methods to interact with the AWS Instance Metadata Service.
+See the documentation for submodules [`fetch_dynamic`], [`fetch_metadata`], and [`fetch_userdata`] for high-level helpers.
```
These aren't really submodules, just methods. :)  Probably safe to just say "methods".

jpculp replied:
I'm going to reword this piece and take out the "submodules" while I'm at it.

zmrow commented on sources/imdsclient/src/lib.rs:
```diff
@@ -0,0 +1,228 @@
+use snafu::ResultExt;
+
+const IMDS_BASE_URI: &str = "http://169.254.169.254";
+const IMDS_PINNED_DATE: &str = "2020-10-27";
```
I'm assuming this is the latest date available as of writing?

jpculp replied:
Sure is!

zmrow commented on sources/imdsclient/src/lib.rs:
````diff
@@ -0,0 +1,228 @@
+    /// # Example
+    ///
+    /// ```
+    /// # use imdsclient::ImdsClient;
````
I'm assuming these doctests will fail if you try to build on anything but an EC2 instance?

If that's the case, let's just move them to the docstring at the top of the lib.

webern replied:
They will run as tests even there. You can tell rust not to run examples though https://stackoverflow.com/a/32429385/2779792

jpculp replied:
Would it be preferable to tell rust not to run the examples, or just remove them entirely?

tjkirch replied:
I'd probably remove them; the functions are pretty simple.

zmrow replied:
I'm on board with either.  Even though this isn't a public crate, I appreciate examples in crates I work with.  They're nice for newcomers/beginners as well.

zmrow commented on sources/imdsclient/src/lib.rs:
```diff
@@ -0,0 +1,228 @@
+/// It obtains a session token when it is first instantiated and is reused between helper functions.
+pub struct ImdsClient {
+    client: Client,
+    session_token: String,
```
I'm a little wary about storing the token without giving the user the option to specify how long it will live.  A user may try to use the client in a process that lasts longer than 60 seconds.

I think a default value of 60 seconds is fine as it covers most of the use cases we have (short quick programs that make a request or three quickly), but I'd like to have an additional `new..`  and/or `fetch...` method that gives the user the opportunity to define the token TTL.

In Rust-land, this could be implemented as an additional method like `ImdsClient::new_with_token_ttl()`.  Under the covers it could call the `fetch_token_with_ttl()` method.  Doing it this way avoids `Option`s in the method signature, which requires a user to pass a `Some` or `None` (which IMHO in this case is not a very friendly API).

tjkirch replied:
Or should we implement automatic token refresh?

zmrow replied:
^ also a good option. :)

zmrow reviewed (commented):
Posting what I have for now, will be back with a few more!

webern commented on sources/imdsclient/src/lib.rs:
```diff
@@ -0,0 +1,228 @@
+                Ok(Some(String::from(response_body.trim_end())))
+            }
+
+            // IMDS returns 404 if no user data is given, or if IMDS is disabled, for example
```
```suggestion
            // IMDS returns 404 if no user data is given, or if IMDS is disabled
```

webern reviewed (commented):
Can tests be written with a mock server? You would need another constructor (which could be private) that takes the URLs. If this constructor were public, you could also use it in your doc example for a passing doc test. You could then implement `Default` for the real clients to use.

jpculp commented:
- Removed examples from docstrings.
- Made the `ImdsClient` struct mutable.
- Added `refresh_token` function. When receiving a 401 response from IMDS, the token will be refreshed, and the IMDS fetch be re-attempted 1 time.
- Tweaked `README` wording.
- Removed extraneous `generate_constants` from build.rs

jpculp commented on sources/imdsclient/src/lib.rs:
```diff
@@ -0,0 +1,210 @@
+        }
+    }
+
+    pub fn refresh_token(&mut self) -> Result<()> {
```
I'll add a docstring to this before I pull this PR out of draft.

jpculp commented:
- Add docstring to `refresh_token`
- Tweaked `ImdsData` error in `shibaken` to be more visually consistent with `ImdsClient` error.

zmrow commented on sources/api/shibaken/src/main.rs:
```diff
@@ -45,75 +38,56 @@ impl UserData {
-        let public_key = public_key_text.trim_end();
+    let target_count: u32 = 0;
+    for target in &public_key_targets {
+        let target_count = target_count + 1;
```
If I'm understanding this correctly, this variable should match up with the index of the key currently being fetched?  If that's the case, rather than keeping a count yourself you can use the [`enumerate()`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.enumerate) method on `public_key_targets` to avoid any "off by one" errors.

zmrow replied:
But perhaps I'm misunderstanding the purpose here.

jpculp replied:
This is just for some extra status context, taking advantage of the description variable that wasn't in the original `shibaken` IMDS logic, but is present in `early-boot-config`. This count is unrelated to the index number (for example if index `0` is the only available key then it will be `public key (1/1)`)

The generated log messages that look something like:
`Requesting public key (1/2) from http://169.254.169.254/latest/meta-data/public-keys/0/openssh-key`

webern commented on sources/imdsclient/src/lib.rs:
```diff
@@ -0,0 +1,212 @@
+}
+
+/// Helper to fetch an IMDSv2 session token that is valid for 60 seconds.
+pub fn fetch_token(client: &Client) -> Result<String> {
```
Is fetch_token actually part of the public interface for this library? I don't see it being used in the clients. If it's not used in the clients maybe we should make it private.

tjkirch replied:
In my mind, ideally, the user doesn't know or care if they're using tokens or IMDSv2 or IMDSv1, unless they ask...

jpculp replied:
It was in the first iteration, but now it's no longer necessary. I'll remove it.

webern commented on sources/imdsclient/src/lib.rs:
```diff
@@ -0,0 +1,212 @@
+    }
+
+    /// Helper to fetch `dynamic` targets from IMDS, preferring an override file if present.
+    pub fn fetch_dynamic(&mut self, end_target: &str, description: &str) -> Result<Option<String>> {
```
Is fetch_dynamic a local helper or something we expect client programs to use?

jpculp replied:
The only case I know of is that `early-boot-config` needs it to [fetch the region](https://github.com/bottlerocket-os/bottlerocket/blob/bf2a78ab8c8e628d8b2c6365bdce3958b0635d4f/sources/api/early-boot-config/src/provider/aws.rs#L149) for the API.

webern replied:
OK, if it's being used then pub is fine. I was just looking for pub lock-down opportunities.

webern commented on sources/imdsclient/src/lib.rs:
```diff
@@ -0,0 +1,212 @@
+
+                // IMDS returns 401 if the session token is expired or invalid
+                StatusCode::UNAUTHORIZED => {
+                    warn!("Session token is invalid or expired");
```
Maybe this is actually `info!` since it's not actually a problem and we are going to do the refresh. Sometimes when you're debugging something hard, warnings can lead you astray.

etungsten reviewed (approved):
LGTM assuming all the existing comments are addressed.

zmrow commented on sources/imdsclient/src/lib.rs:
```diff
@@ -0,0 +1,212 @@
+    }
+
+    /// Helper to fetch `meta-data` targets from IMDS, preferring an override file if present.
+    pub fn fetch_metadata(
```
I'm a bit worried about this API since we're not handling these strings as URLs and we're not doing any handling of slashes, etc.  Ideally we'd at least try to parse the string as a URL so we at least know it's valid.

jpculp replied:
I was going to rely on the `str` -> `Url` parse in `Reqwest`'s `get` function, but it probably makes sense to parse the `uri`  just before the loop.

zmrow commented on sources/imdsclient/src/lib.rs:
```diff
@@ -0,0 +1,212 @@
+    pub fn fetch_imds(
+        &mut self,
+        schema_version: &str,
+        target: &str,
```
Any reason these `&str`s shouldn't be `AsRef<str>`?

jpculp replied:
I didn't think about it, but I think it's a good idea to make that change.

jpculp commented:
> Can tests be written with a mock server? You would need another constructor (which could be private) that takes the URLs. If this constructor were public, you could also use it in your doc example for a passing doc test. You could then implement `Default` for the real clients to use. -@webern

I've gone ahead and added an issue (https://github.com/bottlerocket-os/bottlerocket/issues/1398) for this so that we can revisit it at a later date.

jpculp commented:
- Replaced `&str` with `AsRef<str>` for public functions
- Changed `attempt` from `u16` to `u8`
- Created `max_attempts` variable
- Changed session token expired message from `warn` to `info`
- Tweaked a description string in `shibaken` for a clearer log message.

zmrow reviewed (approved):
:key:

jpculp commented:
Since this is an optimization and there is no rush, I downgraded this to a draft so we can tackle the testing described in [#1398](https://github.com/bottlerocket-os/bottlerocket/issues/1398) and integrate with `pluto` and `early-boot-config` in the same PR.

jpculp commented:
Rebased on v1.1.0

jpculp commented:
Added async functionality in preparation for on-boarding `pluto`.

jpculp commented:
Added a commit on-boarding `pluto`

etungsten commented on sources/api/pluto/src/main.rs:
```diff
@@ -146,23 +125,22 @@ use error::PlutoError;
-        .context(error::ImdsText { uri })
-}
-
-async fn get_max_pods(client: &Client, session_token: &str) -> Result<String> {
-    let instance_type =
-        get_text_from_imds(&client, IMDS_INSTANCE_TYPE_ENDPOINT, session_token).await?;
+    {
+        Ok(Some(instance_type)) => instance_type,
+        Ok(None) => {
+            return error::ImdsData {
+                target: instance_type_target,
+                target_type: "meta-data",
+            }
+            .fail()
+        }
+        Err(e) => Err(e).context(error::ImdsRequest)?,
+    };
```
I'm pretty sure this can be simplified a bit.
Can you try `let instance_type = imdsclient.fetch_metadata(&instance_type_target, "instance type").context(error::ImdsRequest)?.context(error::ImdsData{target : instance_type_target, target_type: "meta-data"})?;`

Same for all other similar constructs in the PR.

etungsten commented on sources/api/pluto/src/main.rs:
```diff
@@ -187,7 +165,7 @@ async fn get_max_pods(client: &Client, session_token: &str) -> Result<String> {
 /// obtained by substituting `10` for the last octet. If the EKS call is not successful, it falls
 /// back to using IMDS MAC CIDR blocks to return one of two default addresses.
-async fn get_cluster_dns_ip(client: &Client, session_token: &str) -> Result<String> {
+async fn get_cluster_dns_ip(imdsclient: &mut ImdsClient) -> Result<String> {
```
Why does this have to be a mutable reference? Is it because we might need to refresh the auth token?

jpculp replied:
Correct

zmrow commented on sources/api/pluto/src/main.rs:
```diff
@@ -146,23 +125,22 @@ use error::PlutoError;
-        .error_for_status()
-        .context(error::ImdsResponse { uri })?
-        .text()
+async fn get_max_pods(imdsclient: &mut ImdsClient) -> Result<String> {
```
Nit: It's typically good Rust style to split words in variables with underscore: `imds_client`

jpculp replied:
What are your thoughts on the variable just being `client`?

jpculp commented:
**imdsclient**
- Replaced output type from `String` to `Vec<u8>` in `fetch_imds`, `fetch_userdata`, and `fetch_dynamic` functions.
- Updated `IMDS_PINNED_DATE` to latest date (2021-01-03).
- Added logic from `early-boot-config` to support compression and shorter log output.

**shibaken & pluto**
- Replaced instances of `imdsclient` with `client`.
- Updated match logic based on @etungsten's suggestion.

**early-boot-config**
- Added async to aws provider and replaced IMDS logic with shared imdsclient library.

jpculp commented:
Fixed and tested `vmware` and `aws-ecs` builds.

jpculp commented:
Added tests to `imdsclient`.

jpculp commented:
- In collaboration with @webern, moved IMDS functions from `shibaken`, `pluto`, and `early-boot-config`, and replaced them with more abstracted functions like `fetch_mac_addresses`.
- Replaced `fetch_metadata` / `fetch_dynamic` with `fetch_string` / `fetch_bytes`.
- Made `refresh_token` private.
- Removed requirement to always add a `description` value.
- @webern added some additional tests for printable strings and parsing public keys.

webern commented on sources/api/pluto/src/main.rs:
```diff
@@ -48,51 +48,26 @@ const DEFAULT_DNS_CLUSTER_IP: &str = "10.100.0.10";
 
-        #[snafu(display("Error getting text response from {}: {}", uri, source))]
-        ImdsText { uri: String, source: reqwest::Error },
+        #[snafu(display("IMDS client failed: Response '404' while fetching '{}'", what))]
```
This is not actually related to 404's anymore.

```suggestion
        #[snafu(display("No '{}' found", what))]
```

webern commented on sources/imdsclient/src/lib.rs:
```diff
@@ -0,0 +1,726 @@
+        #[snafu(display("Response was not UTF-8: {}", source))]
+        NonUtf8Response { source: std::string::FromUtf8Error },
+
+        #[snafu(display("404 file not found fetching {}", uri))]
```
```suggestion
        #[snafu(display("404 file not found fetching '{}'", uri))]
```

webern reviewed (approved):
Nice refactor with new test code. 👍

jpculp commented:
Tweaked the two error messages from @webern's feedback.

bcressey commented on sources/Cargo.lock:
```diff
@@ -2742,12 +2782,13 @@ version = "0.1.0"
 dependencies = [
  "base64",
  "cargo-readme",
+ "imdsclient",
```
This reminds me that we should conditionalize shibaken in `os.spec`, like we do for pluto, since we don't want to run it on non-AWS platforms.

jpculp replied:
That makes sense to me, although I'd want to do that in a separate PR. I'll make an issue so I don't forget. (https://github.com/bottlerocket-os/bottlerocket/issues/1590)

bcressey commented on sources/imdsclient/src/lib.rs:
```diff
@@ -0,0 +1,726 @@
+            target.as_ref()
+        );
+        debug!("Requesting {}", &uri);
+        let mut attempt: u8 = 1;
+        let max_attempts: u8 = 3;
+        loop {
+            attempt += 1;
+            ensure!(attempt <= max_attempts, error::FailedFetch { attempt });
```
Seems like we'll only try twice?

```suggestion
        let mut attempt: u8 = 0;
        let max_attempts: u8 = 3;
        loop {
            attempt += 1;
            ensure!(attempt <= max_attempts, error::FailedFetch { attempt });
```

webern replied:
~~I think it's correct `attempt <= max_attempts`~~

edit: Disregard

bcressey commented on sources/imdsclient/src/lib.rs:
```diff
@@ -0,0 +1,726 @@
+                    return Ok(response_body);
+                }
+
+                // IMDS returns 404 if no user data is given, or if IMDS is disabled
+                StatusCode::NOT_FOUND => return Err(error::Error::NotFound { uri }),
+
+                // IMDS returns 401 if the session token is expired or invalid
+                StatusCode::UNAUTHORIZED => {
+                    info!("Session token is invalid or expired");
+                    self.refresh_token().await?;
+                    info!("Refreshed session token");
+                    continue;
+                }
```
Do we still want to burn one of our retry attempts in this case?

jpculp commented:
**imdsclient**
- Changed number of retries.
- Added 100ms delay between retries.
- Added logic to retry on request timeouts (instead of just expired tokens).

jpculp committed:
shibaken: update to use imdsclient library

This removes the stand-alone IMDS logic in favor of the shared
imdsclient library.

Log messages were also tweaked for more clarity.

jpculp committed:
pluto: update to use imdsclient library

This removes the stand-alone IMDS logic in favor of the shared
imdsclient library.

jpculp committed:
early-boot-config: update to use imdsclient library

This removes the stand-alone IMDS logic in favor of the shared
imdsclient library.

jpculp commented:
**imdsclient**
- Removed extraneous semicolon.
